pkg/resource/deploy/plan.go

```diff
@@ -0,0 +1,612 @@
+
+	for k := range planDiff.Updates {
+		// We expected an update, make sure the value is in the new inputs as an update (not an add)
+		expected := planDiff.Updates[k]
+
+		// If this is in diff.Updates we'll of already checked it
+		_, inUpdates := diff.Updates[k]
+
+		if !inUpdates {
+			// Check if this was in adds, it's not ok to have an update constraint but actually do an add
+			if actual, has := diff.Adds[k]; has {
+				// Constraint wants to update it, but diff wants to add it
+				changes = append(changes, planDiff.MakeError(k, "+", &actual))
+			} else if actual, has := news[k]; has {
+				// It wasn't in the diff as an add so check we have a same
+				if !expected.DeepEqualsIncludeUnknowns(actual) {
+					// diff wants to same this with value X but constraint wants to update with value Y
+					changes = append(changes, planDiff.MakeError(k, "=", &actual))
+				}
+			} else {
+				// Not a same or an update but constraint wants to update it
+
+				// Check if this was <computed> origionally because that could of resolved to undefined
+				// and thus it's ok to be missing, else this is a real missing property
+				if !expected.IsComputed() {
+					changes = append(changes, planDiff.MakeError(k, "-", nil))
+				}
+			}
+		}
+	}
+
+	for i := range planDiff.Deletes {
+		// We expected a delete, make sure its not present
+		k := planDiff.Deletes[i]
+
+		// If this is in diff.Deletes we'll of already checked it
+		_, inDeletes := diff.Deletes[k]
+		if !inDeletes {
+			// See if this is an add, update, or same
+			if actual, has := diff.Adds[k]; has {
+				// Constraint wants to delete this but diff wants to add it
+				changes = append(changes, planDiff.MakeError(k, "+", &actual))
+			} else if actual, has := diff.Updates[k]; has {
+				// Constraint wants to delete this but diff wants to update it
+				changes = append(changes, planDiff.MakeError(k, "~", &actual.New))
+			} else if actual, has := diff.Sames[k]; has {
+				// Constraint wants to delete this but diff wants to leave it same
+				changes = append(changes, planDiff.MakeError(k, "=", &actual))
+			}
+		}
+	}
+
+	if len(changes) > 0 {
+		// Sort changes, mostly so it's easy to write tests against determinstic strings
+		sort.Strings(changes)
+		return fmt.Errorf("properties changed: %v", strings.Join(changes, ", "))
+	}
+
+	return nil
+}
+
+func (rp *ResourcePlan) checkOutputs(
+	oldOutputs resource.PropertyMap,
+	newOutputs resource.PropertyMap,
+) error {
+	contract.Assert(rp.Goal != nil)
+
+	// Check that the property diffs meet the constraints set in the plan.
+	if rp.Goal.OutputDiff != nil {
+		if err := checkDiff(oldOutputs, newOutputs, *rp.Goal.OutputDiff); err != nil {
+			return err
+		}
+	} else {
+		// This plan saved no diff for outputs so new up an empty output diff and diff against that
+		outputDiff := PlanDiff{}
+		if err := checkDiff(oldOutputs, newOutputs, outputDiff); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (rp *ResourcePlan) checkGoal(
```

Comment:
This code is written with early termination semantic (return first error), would it be sometimes helpful to collect and present all errors or it does not matter?

Reply:
Probably, need to think about how best to display large error sets to the user. I think this will be ok for preview release.